demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,439 @@
+
+    /* Iterate through record list to find matching topics, and invoke their callbacks. */
+    while( listIndex < MAX_SUBSCRIPTION_CALLBACK_RECORDS )
+    {
+        if( ( callbackRecordList[ listIndex ].pTopicFilter != NULL ) &&
+            ( matchTopic( pPublishInfo->pTopicName,
+                          pPublishInfo->topicNameLength,
+                          callbackRecordList[ listIndex ].pTopicFilter,
+                          callbackRecordList[ listIndex ].topicFilterLength ) == true ) )
+        {
+            LogInfo( ( "Invoking subscription callback of matching topic filter: "
+                       "TopicFilter=%.*s, TopicName=%.*s",
+                       callbackRecordList[ listIndex ].topicFilterLength,
+                       callbackRecordList[ listIndex ].pTopicFilter,
+                       pPublishInfo->topicNameLength,
+                       pPublishInfo->pTopicName ) );
+
+            /* Invoke the callback associated with the record as the topics match. */
+            callbackRecordList[ listIndex ].callback( pContext, pPublishInfo );
+        }
+
+        listIndex++;
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+bool SubscriptionManager_RegisterCallback( const char * pTopicFilter,
+                                           uint16_t topicFilterLength,
+                                           SubscriptionManager_Callback_t callback )
+{
+    assert( pTopicFilter != NULL );
+    assert( topicFilterLength != 0 );
+
+    bool recordAdded = false;
+    size_t availableIndex = 0u;
+    bool recordExists = false;
+
+    /* Search for an available spot in the list to store the record */
+    while( ( availableIndex < MAX_SUBSCRIPTION_CALLBACK_RECORDS ) &&
+           ( callbackRecordList[ availableIndex ].pTopicFilter != NULL ) )
+    {
+        /* Check if a record for the topic filter already exists in the registry. */
+        if( ( callbackRecordList[ availableIndex ].topicFilterLength == topicFilterLength ) &&
+            ( strncmp( pTopicFilter, callbackRecordList[ availableIndex ].pTopicFilter, topicFilterLength )
+              == 0 ) )
+        {
+            recordExists = true;
+        }
+
+        availableIndex++;
+    }
+
+    if( ( availableIndex == MAX_SUBSCRIPTION_CALLBACK_RECORDS ) )
+    {
+        /* The record list is full. */
+        LogError( ( "Unable to register callback: Registry list is full: TopicFilter=%.*s",
+                    topicFilterLength,
+                    pTopicFilter ) );
+    }
+    else if( recordExists == true )
+    {
+        /* The record for the topic filter already exists. */
+        LogError( ( "Failed to register callback: Record for topic filter already exists: TopicFilter=%.*s",
+                    topicFilterLength,
+                    pTopicFilter ) );
+    }
+    else
+    {
+        /* Should the topic string be copied? */
+        callbackRecordList[ availableIndex ].pTopicFilter = pTopicFilter;
```

Comment:
Is the subscription manager supposed to be specific to one demo? If so, I'm wondering why it is in its own directory. If not, I think this should be changed to a static buffer and memcpy, since if anyone tries to tweak the demo then they'll see issues due to this code that's in a completely different directory

Reply:
The subscription manager is meant to be used by other service client libraries (like Shadow), and is thus, placed in its own folder along with creation of its library target so that demo targets can link against it.
Is your concern that the customer may tweak the demo to use a stack based string pointer instead of a global, and thus, run into memory-out-of scope access problems? 
Would it be fine to mitigate the concern by adding comments in the demo code about keeping the topic filters as global string literals?

Reply:
Yes, we should at least document this somewhere, such as where SubscriptionManager_Record_t is typedef'ed. We should mention the limitation of using a const char * instead of a static buffer

Reply:
Added a `@note` in the documentation of the `RegisterCallback` function